onnxruntime/core/providers/cpu/ml/svmclassifier.cc

```diff
@@ -83,40 +108,48 @@ Status SVMClassifier<T>::Compute(OpKernelContext* ctx) const {
   int64_t N = X->Shape().NumDimensions() == 1 ? 1 : X->Shape()[0];
 
   Tensor* Y = ctx->Output(0, TensorShape({N}));
-  Tensor* Z;
-
   std::vector<int64_t> dims;
-  if (mode_ == SVM_TYPE::SVM_SVC && proba_.size() == 0)
-    dims = {static_cast<int64_t>(N), static_cast<int64_t>(class_count_ * (class_count_ - 1) / 2)};
-  else
-    dims = {static_cast<int64_t>(N), static_cast<int64_t>(class_count_)};
-  Z = ctx->Output(1, TensorShape(dims));
+  int64_t nc = (proba_.size() > 0 || vector_count_ == 0)
+                   ? class_count_
+                   : (class_count_ > 2 ? class_count_ * (class_count_ - 1) / 2 : 2);
+  dims = {static_cast<int64_t>(N), static_cast<int64_t>(nc)};
+  Tensor* Z = ctx->Output(1, TensorShape(dims));
 
-  const auto* x_data = X->template Data<T>();
+  const T* x_data = X->template Data<T>();
   int64_t zindex = 0;
 
   for (int64_t n = 0; n < N; n++)  //for each example
   {
     int64_t current_weight_0 = n * stride;
     int64_t maxclass = -1;
-    double maxweight = 0.f;
     std::vector<float> decisions;
     std::vector<float> scores;
     std::vector<float> kernels;
     std::vector<int64_t> votes;
 
-    if (mode_ == SVM_TYPE::SVM_SVC) {
+    if (vector_count_ == 0 && mode_ == SVM_TYPE::SVM_LINEAR) {
+      // This was in the original code but it does not appear in libsvm or scikit-learn.
+      for (int64_t j = 0; j < class_count_; j++) {  //for each class
+        auto val = kernel_dot(x_data, current_weight_0, coefficients_, feature_count_ * j,
+                              feature_count_, get_kernel_type());
+        val += rho_[0];
+        scores.push_back(val);
+      }
+    } else {
+      if (vector_count_ == 0)
+        return Status(common::ONNXRUNTIME, common::FAIL, "No support vectors.");
+      int evals = 0;
+
       for (int64_t j = 0; j < vector_count_; j++) {
-        float val = kernel_dot(x_data, current_weight_0, support_vectors_, feature_count_ * j, feature_count_, get_kernel_type());
+        auto val = kernel_dot(x_data, current_weight_0, support_vectors_, feature_count_ * j,
+                              feature_count_, get_kernel_type());
         kernels.push_back(val);
       }
-      for (int64_t j = 0; j < class_count_; j++) {
-        votes.push_back(0);
-      }
-      int evals = 0;
-      for (int64_t i = 0; i < class_count_; i++) {        //for each class
-        for (int64_t j = i + 1; j < class_count_; j++) {  //for each class
-          float sum = 0;
+      votes.resize(class_count_, 0);
+      double sum;
```

Comment:
>double sum; [](start = 6, length = 11)

Unitialized var indicated wrong scoping. It has to be declared where first used, in this case in the inner loop. It will also help avoid compiler warnings. If you are not using a var, do not declare it.

Reply:
Fixed. Old habit from C.